stub/storage.go

```diff
@@ -160,45 +162,96 @@ func renderFieldAsString(fields map[string]interface{}) string {
 	return template
 }
 
-func equals(input1, input2 map[string]interface{}) bool {
-	return reflect.DeepEqual(input1, input2)
+func deepEqual(expect, actual interface{}) bool {
+	return reflect.DeepEqual(expect, actual)
+}
+
+func regexMatch(expect, actual interface{}) bool {
+	match, err := regexp.Match(expect.(string), []byte(actual.(string)))
+	if err != nil {
+		log.Printf("Error on matching regex %s with %s error:%v\n", expect, actual, err)
+	}
+	return match
+}
+
+func equals(expect, actual map[string]interface{}) bool {
+	return find(expect, actual, true, true, deepEqual)
 }
 
 func contains(expect, actual map[string]interface{}) bool {
-	for key, val := range expect {
-		actualvalue, ok := actual[key]
-		if !ok {
-			return ok
+	return find(expect, actual, true, false, deepEqual)
+}
+
+func matches(expect, actual map[string]interface{}) bool {
+	return find(expect, actual, true, false, regexMatch)
+}
+
+func find(expect, actual interface{}, acc, exactMatch bool, f matchFunc) bool {
+
+	// circuit brake
+	if acc == false {
+		return false
+	}
+
+	expectArrayValue, expectArrayOk := expect.([]interface{})
+	if expectArrayOk {
+
+		actualArrayValue, actualArrayOk := actual.([]interface{})
+		if !actualArrayOk {
+			acc = false
+			return acc
 		}
 
-		if !reflect.DeepEqual(val, actualvalue) {
-			return false
+		if exactMatch {
+			if len(expectArrayValue) != len(actualArrayValue) {
+				acc = false
+				return acc
+			}
+		} else {
+			if len(expectArrayValue) > len(actualArrayValue) {
+				acc = false
+				return acc
+			}
 		}
-	}
-	return true
-}
 
-func matches(expect, actual map[string]interface{}) bool {
-	for keyExpect, valueExpect := range expect {
-		valueExpectString, ok := valueExpect.(string)
-		if !ok {
-			return false
+		for expectItemIndex, expectItemValue := range expectArrayValue {
+			actualItemValue := actualArrayValue[expectItemIndex]
+			acc = find(expectItemValue, actualItemValue, acc, exactMatch, f)
 		}
-		actualvalue, ok := actual[keyExpect].(string)
-		if !ok {
-			return false
+
+		return acc
+	}
+
+	expectMapValue, expectMapOk := expect.(map[string]interface{})
+	if expectMapOk {
+
+		actualMapValue, actualMapOk := actual.(map[string]interface{})
+		if !actualMapOk {
+			acc = false
+			return acc
 		}
 
-		match, err := regexp.Match(valueExpectString, []byte(actualvalue))
-		if err != nil {
-			log.Printf("Error on matching regex %s with %s error:%v\n", valueExpectString, actualvalue, err)
+		if exactMatch {
+			if len(expectMapValue) != len(actualMapValue) {
+				acc = false
+				return acc
+			}
+		} else {
+			if len(expectMapValue) > len(actualMapValue) {
+				acc = false
+				return acc
+			}
 		}
 
-		if !match {
-			return false
+		for expectItemKey, expectItemValue := range expectMapValue {
+			actualItemValue := actualMapValue[expectItemKey]
+			acc = find(expectItemValue, actualItemValue, acc, exactMatch, f)
 		}
+
+		return acc
 	}
-	return true
+
+	return f(expect, actual)
```

Comment:
If there is no array or JSON object there is nothing to loop over, just do the comparison.